File: packages/google_mlkit_pose_detection/android/src/main/java/com/google_mlkit_pose_detection/PoseDetector.java
package com.google_mlkit_pose_detection;

import android.content.Context;

import androidx.annotation.NonNull;

import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.pose.PoseDetection;
import com.google.mlkit.vision.pose.PoseLandmark;
import com.google.mlkit.vision.pose.accurate.AccuratePoseDetectorOptions;
import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions;
import com.google_mlkit_commons.InputImageConverter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class PoseDetector implements MethodChannel.MethodCallHandler {
    private static final String START = "vision#startPoseDetector";
    private static final String CLOSE = "vision#closePoseDetector";

    private final Context context;
    private final Map<String, com.google.mlkit.vision.pose.PoseDetector> instances = new HashMap<>();

    public PoseDetector(Context context) {
        this.context = context;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method = call.method;
        switch (method) {
            case START:
                handleDetection(call, result);
                break;
            case CLOSE:
                closeDetector(call);
                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void handleDetection(MethodCall call, final MethodChannel.Result result) {
        Map<String, Object> imageData = (Map<String, Object>) call.argument("imageData");
        InputImage inputImage = InputImageConverter.getInputImageFromData(imageData, context, result);
        if (inputImage == null) return;

        String id = call.argument("id");
        com.google.mlkit.vision.pose.PoseDetector poseDetector = instances.get(id);
        if (poseDetector == null) {
            Map<String, Object> options = call.argument("options");
            if (options == null) {
                result.error("PoseDetectorError", "Invalid options", null);
                return;
            }

            String mode = (String) options.get("mode");
            int detectorMode = PoseDetectorOptions.STREAM_MODE;
            if (mode.equals("single")) {
                detectorMode = PoseDetectorOptions.SINGLE_IMAGE_MODE;
            }

            String model = (String) options.get("model");
            if (model.equals("base")) {
                PoseDetectorOptions detectorOptions = new PoseDetectorOptions.Builder()
                        .setDetectorMode(detectorMode)
                        .build();
                poseDetector = PoseDetection.getClient(detectorOptions);
            } else {
                AccuratePoseDetectorOptions detectorOptions = new AccuratePoseDetectorOptions.Builder()
                        .setDetectorMode(detectorMode)
                        .build();
                poseDetector = PoseDetection.getClient(detectorOptions);
            }
            instances.put(id, poseDetector);
        }

        poseDetector.process(inputImage)
                .addOnSuccessListener(
                        pose -> {
                            List<List<Map<String, Object>>> array = new ArrayList<>();
                            if (!pose.getAllPoseLandmarks().isEmpty()) {
                                List<Map<String, Object>> landmarks = new ArrayList<>();
                                for (PoseLandmark poseLandmark : pose.getAllPoseLandmarks()) {
                                    Map<String, Object> landmarkMap = new HashMap<>();
                                    landmarkMap.put("type", poseLandmark.getLandmarkType());
                                    landmarkMap.put("x", poseLandmark.getPosition3D().getX());
                                    landmarkMap.put("y", poseLandmark.getPosition3D().getY());
                                    landmarkMap.put("z", poseLandmark.getPosition3D().getZ());
                                    landmarkMap.put("likelihood", poseLandmark.getInFrameLikelihood());
                                    landmarks.add(landmarkMap);
                                }
                                array.add(landmarks);
                            }
                            result.success(array);
                        })
                .addOnFailureListener(
                        e -> result.error("PoseDetectorError", e.toString(), null));
    }

    private void closeDetector(MethodCall call) {
        String id = call.argument("id");
        com.google.mlkit.vision.pose.PoseDetector poseDetector = instances.get(id);
        if (poseDetector == null) return;
        poseDetector.close();
        instances.remove(id);
    }
}

File: packages/google_mlkit_pose_detection/android/src/main/kotlin/com/google_mlkit_pose_detection/PoseDetector.kt
package com.google_mlkit_pose_detection

import android.content.Context
import com.google.mlkit.vision.pose.PoseDetection
import com.google.mlkit.vision.pose.PoseLandmark
import com.google.mlkit.vision.pose.accurate.AccuratePoseDetectorOptions
import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions
import com.google_mlkit_commons.InputImageConverter
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class PoseDetector(
    private val context: Context,
) : MethodChannel.MethodCallHandler {
    private val instances = mutableMapOf<String, com.google.mlkit.vision.pose.PoseDetector>()

    companion object {
        private const val START = "vision#startPoseDetector"
        private const val CLOSE = "vision#closePoseDetector"
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            START -> {
                handleDetection(call, result)
            }

            CLOSE -> {
                closeDetector(call)
                result.success(null)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun handleDetection(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val imageData =
            call.argument<Map<String, Any>>("imageData") ?: run {
                result.error("PoseDetectorError", "imageData is null", null)
                return
            }
            
        val inputImage = InputImageConverter.getInputImageFromData(imageData, context, result) ?: return

        val id = call.argument<String>("id") ?: return
        var poseDetector = instances[id]

        if (poseDetector == null) {
            val options =
                call.argument<Map<String, Any>>("options") ?: run {
                    result.error("PoseDetectorError", "Invalid options", null)
                    return
                }

            val detectorMode =
                if (options["mode"] as? String == "single") {
                    PoseDetectorOptions.SINGLE_IMAGE_MODE
                } else {
                    PoseDetectorOptions.STREAM_MODE
                }

            poseDetector =
                if (options["model"] as? String == "base") {
                    PoseDetection.getClient(
                        PoseDetectorOptions
                            .Builder()
                            .setDetectorMode(detectorMode)
                            .build(),
                    )
                } else {
                    PoseDetection.getClient(
                        AccuratePoseDetectorOptions
                            .Builder()
                            .setDetectorMode(detectorMode)
                            .build(),
                    )
                }
            instances[id] = poseDetector
        }

        poseDetector
            .process(inputImage)
            .addOnSuccessListener { pose ->
                val array = mutableListOf<List<Map<String, Any>>>()
                if (pose.allPoseLandmarks.isNotEmpty()) {
                    val landmarks =
                        pose.allPoseLandmarks.map { poseLandmark ->
                            mapOf(
                                "type" to poseLandmark.landmarkType,
                                "x" to poseLandmark.position3D.x,
                                "y" to poseLandmark.position3D.y,
                                "z" to poseLandmark.position3D.z,
                                "likelihood" to poseLandmark.inFrameLikelihood,
                            )
                        }
                    array.add(landmarks)
                }
                result.success(array)
            }.addOnFailureListener { e ->
                result.error("PoseDetectorError", e.toString(), null)
            }
    }

    private fun closeDetector(call: MethodCall) {
        val id = call.argument<String>("id") ?: return
        instances.remove(id)?.close()
    }
}
